File: rust_src/input/atmospheric_types.rs
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum CloudType {
    #[default]
    None,
    LambWavelIndep,
    Lambertian,
    MieScattering,
    HgScattering,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum AerosolType {
    #[default]
    None,
    LambWavelIndep,
    Lambertian,
    MieScattering,
    HgScattering,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AbsorberSpecies {
    O2O2,
    O2,
}

impl AbsorberSpecies {
    pub fn is_line_absorbing(self) -> bool {
        matches!(self, Self::O2)
    }

    pub fn is_cross_section(self) -> bool {
        matches!(self, Self::O2O2)
    }

    pub fn is_column_fittable(self) -> bool {
        false
    }

    pub fn is_profile_fittable(self) -> bool {
        self.is_column_fittable()
    }

    pub fn hitran_index(self) -> Option<u8> {
        match self {
            // DISAMAR uses the HITRAN molecule number for line absorption lookup.
            Self::O2 => Some(7),
            Self::O2O2 => None,
        }
    }

    pub fn from_vendor_name(name: &str) -> Option<Self> {
        match name {
            "O2-O2" => Some(Self::O2O2),
            "O2" => Some(Self::O2),
            _ => None,
        }
    }
}

File: src/input/atmospheric_types.zig
const std = @import("std");

pub const AbsorberSpecies = enum {
    o2_o2,
    o2,

    pub fn isLineAbsorbing(self: AbsorberSpecies) bool {
        return switch (self) {
            .o2 => true,
            else => false,
        };
    }

    pub fn isCrossSection(self: AbsorberSpecies) bool {
        return switch (self) {
            .o2_o2 => true,
            else => false,
        };
    }

    pub fn isColumnFittable(self: AbsorberSpecies) bool {
        return switch (self) {
            .o2, .o2_o2 => false,
        };
    }

    pub fn isProfileFittable(self: AbsorberSpecies) bool {
        return self.isColumnFittable();
    }

    pub fn hitranIndex(self: AbsorberSpecies) ?u8 {
        return switch (self) {
            .o2 => 7,
            else => null,
        };
    }

    pub fn fromVendorName(name: []const u8) ?AbsorberSpecies {
        const map = .{
            .{ "O2-O2", .o2_o2 },
            .{ "O2", .o2 },
        };
        inline for (map) |entry| {
            if (std.mem.eql(u8, name, entry[0])) return entry[1];
        }
        return null;
    }
};
